gcloud/datastore/batch.py

```diff
@@ -13,11 +13,55 @@
 # limitations under the License.
 
 """Create / interact with a batch of updates / deletes."""
+try:
+    from threading import local as Local
+except ImportError:     # pragma: NO COVER (who doesn't have it?)
+    class Local(object):
+        """Placeholder for non-threaded applications."""
 
 from gcloud.datastore import _implicit_environ
 from gcloud.datastore import datastore_v1_pb2 as datastore_pb
 
 
+class _Batches(Local):
+    """Manage a thread-local LIFO stack of active batches / transactions.
+
+    Intended for use only in :class:`gcloud.datastore.batch.Batch.__enter__`
+    """
+    def __init__(self):
+        super(_Batches, self).__init__()
+        self._stack = []
+
+    @property
+    def stack(self):
+        """Return a copy of our stack."""
+        return self._stack[:]
+
+    def _push_batch(self, batch):
```

Comment:
This behavior really only works for nested transactions. Any interleaving and the stack is blown.

We don't have to support nesting since it's possible to `put()` and `delete()` into a `Batch` before `commit()`-ing.

I still don't have my head wrapped around what problem this is solving.

I feel like we could get away with much less by replacing the `_Batches` with a single function like `_set_current_context` and then in `__enter__` raising an exception if there is already a context (batch or txn) set.